File: Test/05_model_settings/test_model_settings.py
"""
CASSIA Test 05: Model Settings
==============================
Tests model name resolution, provider shortcuts, and fuzzy alias matching.

Usage:
    python test_model_settings.py
"""

import sys
import time
import io
from pathlib import Path

# Add shared utilities to path
sys.path.insert(0, str(Path(__file__).parent.parent / "shared" / "python"))

from fixtures import get_marker_dataframe_for_cluster
from test_utils import (
    setup_cassia_imports,
    load_config,
    setup_api_keys,
    print_test_header,
    print_test_result,
    print_config_summary
)
from result_manager import (
    create_results_dir,
    save_test_metadata,
    save_test_results,
    create_test_metadata
)

# Setup CASSIA imports
setup_cassia_imports()


def run_model_settings_test():
    """Test model settings and name resolution."""
    print_test_header("05 - Model Settings")

    # Load configuration
    config = load_config()
    print_config_summary(config)

    # Setup API keys
    setup_api_keys()

    # Import CASSIA functions
    from model_settings import resolve_model_name, get_recommended_model, get_available_aliases
    from tools_function import runCASSIA

    # Create results directory
    results_dir = create_results_dir("05_model_settings")
    print(f"Results will be saved to: {results_dir}")

    start_time = time.time()
    errors = []
    test_results = {
        "resolution_tests": [],
        "practical_test": None
    }

    # Test 1: Tier shortcuts
    print("\n" + "="*40)
    print("Testing Tier Shortcuts")
    print("="*40)

    tier_tests = [
        # (tier, provider, expected_contains)
        ("best", "openai", "gpt-5.1"),
        ("balanced", "openai", "gpt-4o"),
        ("fast", "openai", "gpt-5-mini"),
        ("best", "anthropic", "opus"),
        ("balanced", "anthropic", "sonnet"),
        ("fast", "anthropic", "haiku"),
        ("best", "openrouter", "claude"),
        ("fast", "openrouter", "gemini"),
    ]

    tier_passed = 0
    for tier, provider, expected_contains in tier_tests:
        try:
            # Use verbose=False to suppress output for tier tests
            resolved_name, resolved_provider = resolve_model_name(tier, provider, verbose=False)
            success = expected_contains.lower() in resolved_name.lower()
            if not success:
                errors.append(f"Tier '{tier}' resolved to '{resolved_name}', expected to contain '{expected_contains}'")

            test_results["resolution_tests"].append({
                "type": "tier",
                "input": tier,
                "provider": provider,
                "resolved": resolved_name,
                "success": success
            })

            status_symbol = "[OK]" if success else "[X]"
            print(f"  {status_symbol} '{tier}' + '{provider}' -> '{resolved_name}'")

            if success:
                tier_passed += 1

        except Exception as e:
            test_results["resolution_tests"].append({
                "type": "tier",
                "input": tier,
                "provider": provider,
                "error": str(e),
                "success": False
            })
            errors.append(f"Tier resolution failed for '{tier}': {e}")
            print(f"  [X] '{tier}' + '{provider}' -> ERROR: {e}")

    print(f"\nTier tests: {tier_passed}/{len(tier_tests)} passed")

    # Test 2: Fuzzy alias matching
    print("\n" + "="*40)
    print("Testing Fuzzy Alias Matching")
    print("="*40)

    alias_tests = [
        # (alias, provider, expected_contains, should_print_note)
        ("gpt", "openai", "gpt-5.1", True),
        ("gpt", "openrouter", "openai/gpt", True),
        ("claude", "anthropic", "claude-sonnet-4-5", True),
        ("claude", "openrouter", "anthropic/claude-sonnet", True),
        ("gemini", "openrouter", "google/gemini", True),
        ("flash", "openrouter", "gemini-2.5-flash", True),
        ("sonnet", "anthropic", "sonnet", True),
        ("opus", "anthropic", "opus", True),
        ("haiku", "anthropic", "haiku", True),
        ("4o", "openai", "gpt-4o", True),
        ("deepseek", "openrouter", "deepseek", True),
        # Case insensitivity
        ("GPT", "openai", "gpt-5.1", True),
        ("Claude", "anthropic", "claude-sonnet", True),
        # Exact match passthrough (no note)
        ("gpt-4o", "openai", "gpt-4o", False),
        ("claude-sonnet-4-5", "anthropic", "claude-sonnet-4-5", False),
    ]

    alias_passed = 0
    for alias, provider, expected_contains, should_print_note in alias_tests:
        try:
            # Capture stdout to check for notification
            captured = io.StringIO()
            sys.stdout = captured

            resolved_name, _ = resolve_model_name(alias, provider, verbose=True)

            sys.stdout = sys.__stdout__
            output = captured.getvalue()

            # Check result
            result_ok = expected_contains.lower() in resolved_name.lower()

            # Check notification
            note_printed = "Note: Resolved" in output
            notify_ok = note_printed == should_print_note

            success = result_ok and notify_ok

            if not result_ok:
                errors.append(f"Alias '{alias}' resolved to '{resolved_name}', expected to contain '{expected_contains}'")
            if not notify_ok:
                if should_print_note:
                    errors.append(f"Alias '{alias}' should have printed a note but didn't")
                else:
                    errors.append(f"Alias '{alias}' should NOT have printed a note but did")

            test_results["resolution_tests"].append({
                "type": "alias",
                "input": alias,
                "provider": provider,
                "resolved": resolved_name,
                "note_printed": note_printed,
                "success": success
            })

            status_symbol = "[OK]" if success else "[X]"
            note_indicator = "(note)" if note_printed else "(no note)"
            print(f"  {status_symbol} '{alias}' + '{provider}' -> '{resolved_name}' {note_indicator}")

            if success:
                alias_passed += 1

        except Exception as e:
            sys.stdout = sys.__stdout__
            test_results["resolution_tests"].append({
                "type": "alias",
                "input": alias,
                "provider": provider,
                "error": str(e),
                "success": False
            })
            errors.append(f"Alias resolution failed for '{alias}': {e}")
            print(f"  [X] '{alias}' + '{provider}' -> ERROR: {e}")

    print(f"\nAlias tests: {alias_passed}/{len(alias_tests)} passed")

    # Test 3: Provider-aware resolution (same alias, different output)
    print("\n" + "="*40)
    print("Testing Provider-Aware Resolution")
    print("="*40)

    provider_aware_tests = [
        # Same alias resolves differently per provider
        ("claude", "anthropic", "claude-sonnet-4-5"),
        ("claude", "openrouter", "anthropic/claude-sonnet-4.5"),
        ("gpt", "openai", "gpt-5.1"),
        ("gpt", "openrouter", "openai/gpt-5.1"),
    ]

    provider_aware_passed = 0
    for alias, provider, expected in provider_aware_tests:
        try:
            resolved_name, _ = resolve_model_name(alias, provider, verbose=False)
            success = resolved_name == expected

            if not success:
                errors.append(f"Provider-aware: '{alias}' + '{provider}' -> '{resolved_name}', expected '{expected}'")

            status_symbol = "[OK]" if success else "[X]"
            print(f"  {status_symbol} '{alias}' + '{provider}' -> '{resolved_name}' (expected: {expected})")

            if success:
                provider_aware_passed += 1

        except Exception as e:
            errors.append(f"Provider-aware test failed for '{alias}' + '{provider}': {e}")
            print(f"  [X] '{alias}' + '{provider}' -> ERROR: {e}")

    print(f"\nProvider-aware tests: {provider_aware_passed}/{len(provider_aware_tests)} passed")

    # Test 4: Get available aliases
    print("\n" + "="*40)
    print("Testing get_available_aliases()")
    print("="*40)

    try:
        all_aliases = get_available_aliases()
        assert "provider_specific" in all_aliases, "Missing provider_specific"
        assert "global" in all_aliases, "Missing global"
        assert "openrouter" in all_aliases["provider_specific"], "Missing openrouter aliases"
        print("  [OK] get_available_aliases() returns correct structure")

        openai_aliases = get_available_aliases("openai")
        assert "gpt" in openai_aliases["provider_specific"], "Missing 'gpt' alias for openai"
        print("  [OK] get_available_aliases('openai') filters correctly")

        test_results["alias_structure_test"] = {"success": True}
    except Exception as e:
        errors.append(f"get_available_aliases() test failed: {e}")
        print(f"  [X] ERROR: {e}")
        test_results["alias_structure_test"] = {"success": False, "error": str(e)}

    # Calculate total resolution tests
    total_resolution_tests = len(tier_tests) + len(alias_tests) + len(provider_aware_tests)
    total_resolution_passed = tier_passed + alias_passed + provider_aware_passed
    print(f"\nTotal resolution tests: {total_resolution_passed}/{total_resolution_tests} passed")

    # Test 5: Recommended model
    print("\n" + "="*40)
    print("Testing Recommended Model")
    print("="*40)

    try:
        rec_model = get_recommended_model("openrouter")
        print(f"  Recommended for openrouter: {rec_model}")
        test_results["recommended_model"] = rec_model
    except Exception as e:
        errors.append(f"get_recommended_model failed: {e}")
        print(f"  [X] ERROR: {e}")

    # Test 6: Practical test with gemini-2.5-flash via OpenRouter
    print("\n" + "="*40)
    print("Testing Practical Annotation with OpenRouter")
    print("="*40)

    data_config = config['data']
    marker_df = get_marker_dataframe_for_cluster("plasma cell", 20)

    try:
        print(f"  Model: google/gemini-2.5-flash")
        print(f"  Provider: openrouter")
        print(f"  Running annotation...")

        result, _ = runCASSIA(
            model="google/gemini-2.5-flash",
            temperature=0.3,
            marker_list=marker_df,
            tissue=data_config.get('tissue', 'large intestine'),
            species=data_config.get('species', 'human'),
            provider="openrouter",
            validator_involvement="v1"
        )

        practical_success = 'main_cell_type' in result and result['main_cell_type']
        test_results["practical_test"] = {
            "success": practical_success,
            "main_cell_type": result.get('main_cell_type'),
            "sub_cell_types": result.get('sub_cell_types')
        }

        if practical_success:
            print(f"  [OK] Annotation successful")
            print(f"       Main type: {result.get('main_cell_type')}")
        else:
            print(f"  [X] Annotation returned empty result")
            errors.append("Practical test returned empty result")

    except Exception as e:
        test_results["practical_test"] = {
            "success": False,
            "error": str(e)
        }
        errors.append(f"Practical test failed: {e}")
        print(f"  [X] ERROR: {e}")

    duration = time.time() - start_time

    # Determine overall status
    resolution_all_passed = total_resolution_passed == total_resolution_tests
    practical_passed = test_results.get("practical_test", {}).get("success", False)

    # Pass if resolution tests pass OR practical test passes
    if resolution_all_passed and practical_passed:
        status = "passed"
    elif practical_passed:
        status = "passed"  # Practical test is most important
    elif resolution_all_passed:
        status = "passed"  # Resolution tests all passed
    else:
        status = "failed"

    # Save metadata and results
    metadata = create_test_metadata(
        test_name="model_settings",
        config=config,
        duration_seconds=duration,
        status=status,
        clusters_tested=["plasma cell"],
        errors=errors
    )
    save_test_metadata(results_dir, metadata)
    save_test_results(results_dir, test_results)

    # Print final result
    success = status == "passed"
    print_test_result(success, f"Duration: {duration:.2f}s")

    return success


if __name__ == "__main__":
    success = run_model_settings_test()
    sys.exit(0 if success else 1)

File: Test/05_model_settings/test_model_settings.R
# CASSIA Test 05: Model Settings (R)
# ===================================
# Tests model name resolution and provider shortcuts via R package.
#
# Usage:
#     Rscript test_model_settings.R

# Get script directory (works with Rscript and source())
get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg), winslash = "/")))
  }
  for (i in sys.nframe():1) {
    if (!is.null(sys.frame(i)$ofile)) {
      return(dirname(normalizePath(sys.frame(i)$ofile, winslash = "/")))
    }
  }
  return(normalizePath(getwd(), winslash = "/"))
}
script_dir <- get_script_dir()

# Source shared utilities
source(file.path(script_dir, "..", "shared", "r", "test_utils.R"))
source(file.path(script_dir, "..", "shared", "r", "fixtures.R"))
source(file.path(script_dir, "..", "shared", "r", "result_manager.R"))
source(file.path(script_dir, "..", "shared", "r", "logging_manager.R"))

run_model_settings_test <- function() {
  print_test_header("05 - Model Settings (R)")

  # Load configuration
  config <- load_config()
  print_config_summary(config)

  # Setup API keys
  setup_api_keys()

  # Setup CASSIA R package
  tryCatch({
    setup_cassia()
    message("CASSIA R package loaded successfully")
  }, error = function(e) {
    message("Error loading CASSIA: ", e$message)
    return(FALSE)
  })

  # Create results directory
  results <- create_results_dir("05_model_settings", get_test_mode())

  start_logging(results$logs)
  log_msg("Results will be saved to:", results$base)

  start_time <- Sys.time()
  errors <- list()
  test_results <- list(
    resolution_tests = list(),
    practical_test = NULL
  )

  # Test 1: Model name resolution via Python (through reticulate)
  log_separator()
  log_msg("Testing Model Name Resolution")
  log_separator()

  resolution_tests <- list(
    list(simple_name = "gemini", provider = "openrouter", expected_contains = "gemini"),
    list(simple_name = "best", provider = "openrouter", expected_contains = "gemini"),
    list(simple_name = "cheap", provider = "openrouter", expected_contains = NULL),
    list(simple_name = "flash", provider = "openrouter", expected_contains = "flash")
  )

  resolution_passed <- 0

  tryCatch({
    # Import model_settings module via reticulate
    model_settings <- reticulate::import("CASSIA.core.model_settings")

    for (test in resolution_tests) {
      tryCatch({
        resolved <- model_settings$resolve_model_name(test$simple_name, test$provider)
        resolved_name <- resolved[[1]]

        success <- TRUE
        if (!is.null(test$expected_contains)) {
          if (!grepl(tolower(test$expected_contains), tolower(resolved_name))) {
            success <- FALSE
            errors <- c(errors, paste0("'", test$simple_name, "' resolved to '",
                                       resolved_name, "', expected to contain '",
                                       test$expected_contains, "'"))
          }
        }

        test_results$resolution_tests <- c(test_results$resolution_tests, list(list(
          input = test$simple_name,
          provider = test$provider,
          resolved = resolved_name,
          success = success
        )))

        status_symbol <- if (success) "[OK]" else "[X]"
        log_msg("  ", status_symbol, " '", test$simple_name, "' + '", test$provider, "' -> '", resolved_name, "'", sep = "")

        if (success) resolution_passed <- resolution_passed + 1

      }, error = function(e) {
        test_results$resolution_tests <- c(test_results$resolution_tests, list(list(
          input = test$simple_name,
          provider = test$provider,
          error = e$message,
          success = FALSE
        )))
        errors <<- c(errors, paste0("Resolution failed for '", test$simple_name, "': ", e$message))
        log_msg("  [X] '", test$simple_name, "' + '", test$provider, "' -> ERROR: ", e$message, "", sep = "")
      })
    }
  }, error = function(e) {
    log_msg("  [X] Could not import model_settings module:", e$message)
    errors <<- c(errors, paste0("Module import failed: ", e$message))
  })

  log_msg("\nResolution tests:", resolution_passed, "/", length(resolution_tests), "passed")

  # Test 2: Practical test with gemini-2.5-flash via OpenRouter
  log_separator()
  log_msg("Testing Practical Annotation with OpenRouter")
  log_separator()

  data_config <- config$data
  marker_df <- get_marker_dataframe_for_cluster("plasma cell", 20)

  tryCatch({
    log_msg("  Model: google/gemini-2.5-flash")
    log_msg("  Provider: openrouter")
    log_msg("  Running annotation...")

    result <- CASSIA::runCASSIA(
      model = "google/gemini-2.5-flash",
      temperature = 0.3,
      marker_list = marker_df,
      tissue = data_config$tissue %||% "large intestine",
      species = data_config$species %||% "human",
      provider = "openrouter",
      validator_involvement = "v1"
    )

    practical_success <- !is.null(result$main_cell_type) && result$main_cell_type != ""
    test_results$practical_test <- list(
      success = practical_success,
      main_cell_type = result$main_cell_type,
      sub_cell_types = result$sub_cell_types
    )

    if (practical_success) {
      log_msg("  [OK] Annotation successful")
      log_msg("       Main type:", result$main_cell_type)
    } else {
      log_msg("  [X] Annotation returned empty result")
      errors <- c(errors, "Practical test returned empty result")
    }

  }, error = function(e) {
    test_results$practical_test <<- list(
      success = FALSE,
      error = e$message
    )
    errors <<- c(errors, paste0("Practical test failed: ", e$message))
    log_msg("  [X] ERROR:", e$message)
  })

  duration <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

  # Determine overall status
  resolution_all_passed <- resolution_passed == length(resolution_tests)
  practical_passed <- !is.null(test_results$practical_test) &&
                      test_results$practical_test$success

  if (resolution_all_passed && practical_passed) {
    status <- "passed"
  } else if (practical_passed) {
    status <- "passed"  # Practical test is most important
  } else {
    status <- "failed"
  }

  # Save metadata and results
  metadata <- create_test_metadata(
    test_name = "model_settings",
    config = config,
    duration_seconds = duration,
    status = status,
    clusters_tested = list("plasma cell"),
    errors = as.list(errors)
  )
  save_test_metadata(results$outputs, metadata)
  save_test_results(results$outputs, test_results)

  # Print final result
  success <- status == "passed"
  print_test_result(success, paste("Duration:", round(duration, 2), "s"))


  stop_logging()
  return(success)
}

# Run test
success <- run_model_settings_test()
quit(status = if (success) 0 else 1)
